spark3/src/main/java/org/apache/iceberg/actions/Spark3MigrateAction.java

```diff
@@ -0,0 +1,155 @@
+
+/**
+ * Takes a Spark table in the sourceCatalog and attempts to transform it into an Iceberg
+ * Table in the same location with the same identifier. Once complete the identifier which
+ * previously referred to a non-iceberg table will refer to the newly migrated iceberg
+ * table.
+ */
+class Spark3MigrateAction extends Spark3CreateAction {
+  private static final Logger LOG = LoggerFactory.getLogger(Spark3MigrateAction.class);
+  private static final String BACKUP_SUFFIX = "_BACKUP_";
+
+  Spark3MigrateAction(SparkSession spark, CatalogPlugin sourceCatalog, Identifier sourceTableName) {
+    super(spark, sourceCatalog, sourceTableName, sourceCatalog, sourceTableName);
+  }
+
+  @Override
+  public Long execute() {
+    Table icebergTable;
+
+    // Move source table to a new name, halting all modifications and allowing us to stage
+    // the creation of a new Iceberg table in its place
+    String backupName = sourceTableName().name() + BACKUP_SUFFIX;
+    Identifier backupIdentifier = Identifier.of(sourceTableName().namespace(), backupName);
+    try {
+      destCatalog().renameTable(sourceTableName(), backupIdentifier);
+    } catch (NoSuchTableException e) {
+      throw new IllegalArgumentException("Cannot find table to migrate", e);
+    } catch (TableAlreadyExistsException e) {
+      throw new IllegalArgumentException("Cannot rename migration source to backup name", e);
+    }
+
+    StagedTable stagedTable = null;
+    boolean threw = true;
+    try {
+      Map<String, String> newTableProperties = new HashMap<>();
+      newTableProperties.put(TableCatalog.PROP_PROVIDER, "iceberg");
+      newTableProperties.putAll(JavaConverters.mapAsJavaMapConverter(v1SourceTable().properties()).asJava());
+      newTableProperties.putAll(tableLocationProperties(sourceTableLocation()));
+      newTableProperties.putAll(additionalProperties());
+
+      stagedTable = destCatalog().stageCreate(Identifier.of(
+          destTableName().namespace(),
+          destTableName().name()),
+          v1SourceTable().schema(),
+          sourcePartitionSpec(),
+          newTableProperties);
+
+      icebergTable = ((SparkTable) stagedTable).table();
+
+      if (!stagedTable.properties().containsKey(TableProperties.DEFAULT_NAME_MAPPING)) {
+        assignDefaultTableNameMapping(icebergTable);
+      }
+
+      String stagingLocation = stagedTable.properties().get(TableProperties.WRITE_METADATA_LOCATION);
+
+      LOG.info("Beginning migration of {} using metadata location {}", sourceTableName(), stagingLocation);
+
+      Some<String> backupNamespace = Some.apply(backupIdentifier.namespace()[0]);
+      TableIdentifier v1BackupIdentifier = new TableIdentifier(backupIdentifier.name(), backupNamespace);
+      SparkTableUtil.importSparkTable(spark(), v1BackupIdentifier, icebergTable, stagingLocation);
+
+      stagedTable.commitStagedChanges();
+      threw = false;
+
+    } catch (TableAlreadyExistsException tableAlreadyExistsException) {
```

Comment:
If we move staging to the parent class, this would also be simplified. It is a bit hard to follow now.

Reply:
This will make the errors much more difficult to reason about. Since in "Snapshot" we will report "table already exists" which might make sense, but in "migrate" it makes no sense since our migration has no destination table